File: app/src/main/java/com/robifr/ledger/ui/searchproduct/recycler/SearchProductAdapter.java
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.searchproduct.recycler;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import com.robifr.ledger.data.model.ProductModel;
import com.robifr.ledger.databinding.ListableListTextBinding;
import com.robifr.ledger.databinding.ProductCardWideBinding;
import com.robifr.ledger.ui.RecyclerViewHolder;
import com.robifr.ledger.ui.product.ProductListAction;
import com.robifr.ledger.ui.searchproduct.SearchProductCardAction;
import com.robifr.ledger.ui.searchproduct.SearchProductFragment;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class SearchProductAdapter extends RecyclerView.Adapter<RecyclerViewHolder<?, ?>>
    implements ProductListAction, SearchProductCardAction {
  private enum ViewType {
    HEADER(0),
    LIST(1);

    private final int _value;

    private ViewType(int value) {
      this._value = value;
    }

    public int value() {
      return this._value;
    }
  }

  @NonNull private final SearchProductFragment _fragment;

  public SearchProductAdapter(@NonNull SearchProductFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);
  }

  @Override
  @NonNull
  public RecyclerViewHolder<?, ?> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    Objects.requireNonNull(parent);

    final ViewType type =
        Arrays.stream(ViewType.values())
            .filter(e -> e.value() == viewType)
            .findFirst()
            .orElse(ViewType.LIST);
    final LayoutInflater inflater = this._fragment.getLayoutInflater();

    return switch (type) {
      case HEADER ->
          new SearchProductHeaderHolder<>(
              ListableListTextBinding.inflate(inflater, parent, false), this);

        // Defaults to `ViewType#LIST`.
      default ->
          new SearchProductListHolder<>(
              ProductCardWideBinding.inflate(this._fragment.getLayoutInflater(), parent, false),
              this);
    };
  }

  @Override
  public void onBindViewHolder(@NonNull RecyclerViewHolder holder, int index) {
    Objects.requireNonNull(holder);

    if (holder instanceof SearchProductHeaderHolder<?> headerHolder) {
      headerHolder.bind(Optional.empty());

    } else if (holder instanceof SearchProductListHolder<?> listHolder) {
      this._fragment
          .searchProductViewModel()
          .products()
          .getValue()
          .map(products -> products.get(index - 1)) // -1 offset because header holder.
          .ifPresent(listHolder::bind);
    }
  }

  @Override
  public int getItemCount() {
    // +1 offset because header holder.
    return this._fragment.searchProductViewModel().products().getValue().map(List::size).orElse(0)
        + 1;
  }

  @Override
  public int getItemViewType(int index) {
    return switch (index) {
      case 0 -> ViewType.HEADER.value();
      default -> ViewType.LIST.value();
    };
  }

  @Override
  @NonNull
  public List<ProductModel> products() {
    return this._fragment.searchProductViewModel().products().getValue().orElse(List.of());
  }

  @Override
  public void onProductSelected(@Nullable ProductModel product) {
    this._fragment.searchProductViewModel().onProductSelected(product);
  }
}

File: app/src/main/java/com/robifr/ledger/ui/searchproduct/recycler/SearchProductAdapter.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.searchproduct.recycler

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.robifr.ledger.databinding.ListableListTextBinding
import com.robifr.ledger.databinding.ProductCardWideBinding
import com.robifr.ledger.ui.RecyclerViewHolderKt
import com.robifr.ledger.ui.product.recycler.ProductListHolder
import com.robifr.ledger.ui.searchproduct.SearchProductFragment
import com.robifr.ledger.ui.selectproduct.recycler.SelectProductListHolder

class SearchProductAdapter(private val _fragment: SearchProductFragment) :
    RecyclerView.Adapter<RecyclerViewHolderKt>() {
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolderKt =
      when (ViewType.entries.find { it.value == viewType }) {
        ViewType.HEADER ->
            SearchProductHeaderHolder(
                _textBinding =
                    ListableListTextBinding.inflate(_fragment.layoutInflater, parent, false),
                _products = { _fragment.searchProductViewModel.uiState.safeValue.products })
        else -> {
          val cardBinding: ProductCardWideBinding =
              ProductCardWideBinding.inflate(_fragment.layoutInflater, parent, false)
          if (_fragment.searchProductViewModel.uiState.safeValue.isSelectionEnabled) {
            SelectProductListHolder(
                _cardBinding = cardBinding,
                _initialSelectedProductIds = {
                  _fragment.searchProductViewModel.uiState.safeValue.initialSelectedProductIds
                },
                _products = { _fragment.searchProductViewModel.uiState.safeValue.products },
                _onProductSelected = _fragment.searchProductViewModel::onProductSelected,
                _expandedProductIndex = {
                  _fragment.searchProductViewModel.uiState.safeValue.expandedProductIndex
                },
                _onExpandedProductIndexChanged =
                    _fragment.searchProductViewModel::onExpandedProductIndexChanged)
          } else {
            ProductListHolder(
                _cardBinding = cardBinding,
                _products = { _fragment.searchProductViewModel.uiState.safeValue.products },
                _onDeleteProduct = _fragment.searchProductViewModel::onDeleteProduct,
                _expandedProductIndex = {
                  _fragment.searchProductViewModel.uiState.safeValue.expandedProductIndex
                },
                _onExpandedProductIndexChanged =
                    _fragment.searchProductViewModel::onExpandedProductIndexChanged)
          }
        }
      }

  override fun onBindViewHolder(holder: RecyclerViewHolderKt, position: Int) {
    when (holder) {
      is SearchProductHeaderHolder -> holder.bind()
      is ProductListHolder -> holder.bind(position - 1)
      is SelectProductListHolder -> holder.bind(position - 1)
    }
  }

  override fun getItemCount(): Int =
      // +1 offset because header holder.
      _fragment.searchProductViewModel.uiState.safeValue.products.size + 1

  override fun getItemViewType(position: Int): Int =
      when (position) {
        0 -> ViewType.HEADER.value
        else -> ViewType.LIST.value
      }

  private enum class ViewType(val value: Int) {
    HEADER(0),
    LIST(1)
  }
}
